### PR TITLE
Feature/v4l usb camera

### PR DESCRIPTION
**Pull request by ezavesky**
* adding early usb camera (via v4l) support to ssl-vision
   * Note: currently lacks ability to set alternate frame resolution and capture rate
* attempts to satisfy some open issues like #33 and #28, maybe even issue #11 (but it does not include OpenCV hooks)
* also sneaks in a few updates to the main README.md file to reflect new usb functionality

### CODE REVIEW

File: src/app/capture_thread.cpp [hunk 1]
```diff
@@ -184,39 +190,40 @@ void CaptureThread::run() {
...
               capture_mutex.unlock();
-            }
-            stack_mutex.lock();
-            stack->updateTimingStatistics();
-            stack_mutex.unlock();
-          }
-          capture_mutex.lock();
-          if ((capture != 0) && (capture->isCapturing())) {
-            capture->releaseFrame();
           }
```
Comment (joydeep-b) [inline src/app/capture_thread.cpp]: Should have some error statement else?
i.e. if bSuccess == false

File: src/app/gui/glLUTwidget.cpp [hunk 2]
```diff
@@ -781,7 +783,8 @@ void GLLUTWidget::glDrawSlice(Slice * s) {
     }
     glEnable(GL_COLOR_LOGIC_OP);
     //glLogicOp(GL_COPY_INVERTED);
-    glLogicOp(GL_XOR);
+    //glLogicOp(GL_XOR);
+    glLogicOp(GL_EQUIV);                //fix drawing inverted color (2/2/16)
```
Comment (joydeep-b) [inline src/app/gui/glLUTwidget.cpp]: What is this for?

File: src/shared/CMakeLists.txt.inc [hunk 3]
```diff
@@ -12,10 +12,11 @@ include_directories(${shared_dir}/vartypes/enhanced)
 include_directories(${shared_dir}/vartypes/gui)
 include_directories(${shared_dir}/vartypes/xml)
 
-include_directories("/usr/include/eigen2")
+# include_directories("/usr/include/eigen2")  # update for eigen3 at toplevel 01/16
```
Comment (joydeep-b) [inline src/shared/CMakeLists.txt.inc]: Is this comment correct? If not, please remove.

File: src/shared/capture/capturev4l.h [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+#include <stdio.h>
+#include <stdlib.h>
+#include <string>
+#include "VarTypes.h"
+#include <linux/videodev2.h>
+#include <sys/poll.h>
+
+#include <map>
+
+//#include "conversions.h"
```
Comment (joydeep-b) [inline src/shared/capture/capturev4l.h]: Please remove.

File: src/app/gui/glLUTwidget.cpp (hunk 2 above)
Comment (ezavesky) [inline src/app/gui/glLUTwidget.cpp]: not sure why the original inversion got commented out, but this line restores the ability to see the LUT marked up with a sample drawn from the active image (now they are visualized with small white 'x's as I think was originally intended)

File: src/app/capture_thread.cpp (hunk 1 above)
Comment (ezavesky) [inline src/app/capture_thread.cpp]: Unfortunately, the auto re-indent caused some mess here, but the main reason is that on a good frame we'll proceed for the whole frame/update but we skip the process on a bad frame.  At the time there was some race condition (e.g. when starting with '-s') that caused a crash with some slow-to-start drivers (like v4l).

File: src/shared/util/rawimage.cpp [hunk 5]
```diff
@@ -150,20 +150,13 @@ void RawImage::clear()
...
-    case COLOR_MONO16:
-    return pixelCount*2;
+    case COLOR_RGB8: return pixelCount*3;
+    case COLOR_RGBA8: return pixelCount*4;
+    case COLOR_YUV444: return pixelCount*3;
+    case COLOR_YUV422_UYVY: return pixelCount*2;
+    case COLOR_YUV411: return pixelCount*3/2;
+    case COLOR_MONO8: return pixelCount;
+    case COLOR_MONO16: return pixelCount*2;
     default:
```
Comment (joydeep-b) [inline src/shared/util/rawimage.cpp]: If changing the previous cases, please change this to be consistent, as well.

File: src/shared/util/rawimage.h [hunk 6]
```diff
@@ -38,7 +38,7 @@
 */
 class RawImage : public ImageInterface
 {
-  protected:
+protected:
```
Comment (joydeep-b) [inline src/shared/util/rawimage.h]: Please revert.

File: src/shared/util/rawimage.h [hunk 7]
```diff
@@ -54,7 +54,7 @@ class RawImage : public ImageInterface
   /// capture timestamp of the image
   double   time;
 
-  public:
+public:
```
Comment (joydeep-b) [inline src/shared/util/rawimage.h]: Please revert.

File: src/app/capture_thread.cpp [hunk 8]
```diff
@@ -36,9 +36,11 @@ CaptureThread::CaptureThread(int cam_id)
   control->addChild( (VarType*) (captureModule= new VarStringEnum("Capture Module","DC 1394")));
   captureModule->addFlags(VARTYPE_FLAG_NOLOAD_ENUM_CHILDREN);
   captureModule->addItem("DC 1394");
+  captureModule->addItem("V4L (USB)");
```
Comment (joydeep-b) [inline src/app/capture_thread.cpp]: Please call this "Video 4 Linux", to be precise. PtGrey and other manufacturers have their own custom USB interface.

File: src/shared/CMakeLists.txt.inc (hunk 3 above)
Comment (ezavesky) [inline src/shared/CMakeLists.txt.inc]: Yes, on some lesser devices (e.g. raspberry-pi ubuntu) there was some difficulty getting the eigen2 library, so the cmake dependency and single line of code were updated. (see 816042cf8d9c5f95f77a82c589c2b12c8772a62c)

**Comment by joydeep-b** [comment]
Thanks for the pull-request, Eric. I reviewed it, and it looks pretty good, but I had some comments (above). Could you please update based on the comments? Thanks.

File: src/shared/capture/capturev4l.h (hunk 4 above)
Comment (joydeep-b) [inline src/shared/capture/capturev4l.h]: (I meant the commented out line)

**Comment by ezavesky** [comment]
Heh, looks like some confusion with the first commit, so we'll aim for something more graceful next time.  I was going to pull request from an isolated branch so that other changes could be pushed, but I can resolve this first (and keep this history) if you prefer.

File: src/shared/capture/capturev4l.h (hunk 4 above)
Comment (ezavesky) [inline src/shared/capture/capturev4l.h]: Not sure what is incorrect about this commit.  Was it the commented-out "conversions.h" include?

**Comment by joydeep-b** [comment]
Can you edit the code based on the comments inline here? Then that'll ensure that the requested edits happen before pulling these changes.

File: src/shared/util/rawimage.cpp (hunk 5 above)
Comment (ezavesky) [inline src/shared/util/rawimage.cpp]: No functional change, just over-zealous readability.

File: src/shared/CMakeLists.txt.inc (hunk 3 above)
Comment (joydeep-b) [inline src/shared/CMakeLists.txt.inc]: I see. Okay, given that the primary platform (Ubuntu LTS) has Eigen3, then we could just remove this commented out line entirely.

File: src/app/capture_thread.cpp (hunk 8 above)
Comment (ezavesky) [inline src/app/capture_thread.cpp]: Fixed in ad663fcf87c382778415e1955a3ad7f265a910f1

File: src/shared/util/rawimage.h (hunk 6 above)
Comment (ezavesky) [inline src/shared/util/rawimage.h]: Fixed in ad663fcf87c382778415e1955a3ad7f265a910f1

File: src/shared/util/rawimage.h (hunk 7 above)
Comment (ezavesky) [inline src/shared/util/rawimage.h]: Fixed in ad663fcf87c382778415e1955a3ad7f265a910f1

File: src/shared/CMakeLists.txt.inc (hunk 3 above)
Comment (ezavesky) [inline src/shared/CMakeLists.txt.inc]: Updated in change 89e70a6e2b6cebfd6f133d7e1e897600c2e8a85e (currently in testing)

**Comment by ezavesky** [comment]
@joydeep-b Okay, all issues have been addressed, please pull at your convenience.   I have another fix/feature for the generator that I think would be useful but can't push because this PR is open on the master branch.

**Comment by joydeep-b** [comment]
LGTM.

**Comment by MichaelEischer** [comment]
Nice to see some new features for SSL-Vision.
I've noticed some comments containing a date and reason why something was changed. For example "//fix drawing inverted color (2/2/16)". along with leaving some old dead code as comments in place. This kind of notes is, however, best managed by the source code management tools. So in my opinion this should be avoid unless there is a good reason for it. `git blame` `git log` and `git revert` (just in case the change wasn't that good after all) would be my preferred tools to replace those comments. (Sorry for nit-picking)
P.S. GitHub managed to send me 28 emails for this issue... ^^

**Pull request merged**